uvicorn/protocols/websockets/wsproto_impl.py

```diff
@@ -254,49 +262,110 @@ async def send(self, message: "ASGISendEvent") -> None:
+                            wsproto.events.AcceptConnection(
+                                subprotocol=subprotocol,
+                                extensions=extensions,
+                                extra_headers=extra_headers,
+                            )
                         )
+                        self.transport.write(output)
+
+                elif message_type == "websocket.close":
+                    self.queue.put_nowait(
+                        {"type": "websocket.disconnect", "code": 1006}
                     )
+                    self.logger.info(
+                        '%s - "WebSocket %s" 403',
+                        self.scope["client"],
+                        get_path_with_query_string(self.scope),
+                    )
+                    self.handshake_complete = True
+                    self.close_sent = True
+                    event = events.RejectConnection(status_code=403, headers=[])
+                    output = self.conn.send(event)
                     self.transport.write(output)
+                    self.transport.close()
 
-            elif message_type == "websocket.close":
-                self.queue.put_nowait({"type": "websocket.disconnect", "code": 1006})
-                self.logger.info(
-                    '%s - "WebSocket %s" 403',
-                    self.scope["client"],
-                    get_path_with_query_string(self.scope),
-                )
-                self.handshake_complete = True
-                self.close_sent = True
-                event = events.RejectConnection(status_code=403, headers=[])
-                output = self.conn.send(event)
-                self.transport.write(output)
-                self.transport.close()
-
+                elif message_type == "websocket.http.response.start":
+                    message = typing.cast("WebSocketResponseStartEvent", message)
+                    # ensure status code is in the valid range
+                    if not (100 <= message["status"] < 600):
+                        msg = "Invalid HTTP status code '%d' in response."
+                        raise RuntimeError(msg % message["status"])
+                    self.logger.info(
+                        '%s - "WebSocket %s" %d',
+                        self.scope["client"],
+                        get_path_with_query_string(self.scope),
+                        message["status"],
+                    )
+                    event = events.RejectConnection(
+                        status_code=message["status"],
+                        headers=list(message["headers"]),
+                        has_body=True,
+                    )
+                    # Create the event here but do not send it, the ASGI spec
+                    # suggest that we wait for the body event before sending.
+                    # https://asgi.readthedocs.io/en/latest/specs/www.html#response-start-send-event
+                    self.reject_event = event
```

Comment:
I know.  I disagree, I think the _recommendation_  is a good one because it enlarges the window during which the server can still send a 500 error in case the application messes up.  It is not a requirement but (IMHO) good practice in a library, and that is the reason I wrote the code that way.
But it is your call of course, happy to get this merged either way.

Reply:
I understand, and it makes sense... But I'll rather not have a mismatch in how the HTTP implementations behave, and the WebSockets.